File: Java/프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
class Solution {
    public String solution(String[] survey, int[] choices) {
        String answer = "";
        String[] type = new String[]{"R", "T", "C", "F", "J", "M", "A", "N"};
        int[] mbti = new int[8];
        int nagIdx = -1;
        int posIdx = -1;
        for(int i=0; i< survey.length; i++) {
            switch(survey[i]) {
                case "RT":
                    nagIdx = 0;
                    posIdx = 1;
                    break;
                case "TR":
                    nagIdx = 1;
                    posIdx = 0;
                    break;
                case "CF":
                    nagIdx = 2;
                    posIdx = 3;
                    break;
                case "FC":
                    nagIdx = 3;
                    posIdx = 2;
                    break;
                case "JM":
                    nagIdx = 4;
                    posIdx = 5;
                    break;
                case "MJ":
                    nagIdx = 5;
                    posIdx = 4;
                    break;
                case "AN":
                    nagIdx = 6;
                    posIdx = 7;
                    break;
                case "NA":
                    nagIdx = 7;
                    posIdx = 6;
                    break;
            }
            switch(choices[i]) {
                case 1:
                    mbti[nagIdx] +=3;
                    break;
                case 2:
                    mbti[nagIdx] +=2;
                    break;
                case 3:
                    mbti[nagIdx] +=1;
                    break;
                case 4:
                    break;
                case 5:
                    mbti[posIdx] +=1;
                    break;
                case 6:
                    mbti[posIdx] +=2;
                    break;
                case 7:
                    mbti[posIdx] +=3;
                    break;
            }
        }
        for(int i = 0; i<4; i++) {
            String result = mbti[2*i] >= mbti[2*i+1] ? type[2*i] : type[2*i+1];
            answer+=result;
        }
        return answer;
    }
}

File: Kotlin/프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.kt
class Solution {
    fun solution(survey: Array<String>, choices: IntArray): String {
        var mbti = arrayOf("R","T","C","F","J","M","A","N")
        var score = Array(8){0}
        var posIndex = -1
        var nagIndex = -1
        survey.forEachIndexed{
            i, v ->
            when(v) {
                "RT" -> {
                    posIndex = 0
                    nagIndex = 1
                }
                "TR" -> {
                    posIndex = 1
                    nagIndex = 0
                }
                "CF" -> {
                    posIndex = 2
                    nagIndex = 3
                }
                "FC" -> {
                    posIndex = 3
                    nagIndex = 2
                }
                "JM" -> {
                    posIndex = 4
                    nagIndex = 5
                }
                "MJ" -> {
                    posIndex = 5
                    nagIndex = 4
                }
                "AN" -> {
                    posIndex = 6
                    nagIndex = 7
                }
                "NA" -> {
                    posIndex = 7
                    nagIndex = 6
                }
                else ->{
                    posIndex = -1
                    nagIndex = -1
                }
            }
           when(choices[i]) {
                   1 -> score[posIndex]+=3
                   2 -> score[posIndex]+=2
                   3 -> score[posIndex]+=1
                   5 -> score[nagIndex]+=1
                   6 -> score[nagIndex]+=2
                   7 -> score[nagIndex]+=3
                   else -> score[posIndex]+=0
               }
        }
        return (0..3).toList().map{
            if(score[2*it] >= score[2*it+1]) mbti[2*it] else mbti[2*it+1]
        }.joinToString("")
    }
}